Prototype of initial feature service for registering features

thecoop opened the pull request:
Add a FeatureService that is responsible for registering & publishing features. Store local features in DiscoveryNode

thecoop committed:
Add initial feature service for registering features

thecoop commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -291,6 +292,13 @@ protected List<LoadedPlugin> plugins() {
         return this.plugins;
     }
 
+    public void registerPluginFeatures(FeatureService service) {
```
I don't like this being here, but the ES version in the descriptor is not accessible outside this class

ldematte replied:
I kind of like it instead. Are you worried about something in particular? e.g. inititialization/timing issues?

ldematte replied:
I also like how registerFeatures is implemented. Easy to use.

elasticsearchmachine commented:
Hi @thecoop, I've created a changelog YAML for you.

thecoop committed:
Update docs/changelog/99944.yaml

thecoop commented on server/src/main/java/org/elasticsearch/features/FeatureService.java:
```diff
@@ -0,0 +1,47 @@
+    private record Feature(String id, int era) implements NodeFeature {}
+
+    public NodeFeature registerFeature(String id, int era) {
+        if (locked) throw new IllegalStateException("The node's feature set has already been read");
```
All feature registration should be done within the `Node` constructor, the set is locked when the node is actually started

thecoop committed:
spotless

thecoop committed:
Delete docs/changelog/99944.yaml

thecoop committed:
Add features to DiscoveryNode

thecoop committed:
Tweak names

ldematte commented on server/src/main/java/org/elasticsearch/features/FeatureService.java:
```diff
@@ -0,0 +1,86 @@
+            // check the sizes to ensure we haven't got dups
+            int sizeBefore = aggregatedFeatures.size();
+            aggregatedFeatures.addAll(versions.getValue());
+            if (aggregatedFeatures.size() < sizeBefore + versions.getValue().size()) {
```
I think addAll on a Set may already remove duplicates, but need to check

ldematte replied:
Oh I see, you are leveraging that to actually detect this condition. 
Why do we care?

thecoop replied:
To check if there are any duplicates in the historical versions list above (which is incorrect)

ldematte commented:
Overall I like it; I have one major concern. 
We are exposing features as a pure Set<String>, correct? So feature detection will be a contains(). I think a level of abstraction here would be nice, e.g. passing down a `FeatureProvider` or similar.

One place where this might be immediately good is for HISTORICAL_FEATURES; we might want to always answer "true" to these even if they are eras before, while still trimming them/remove them from the exchanged feature set (via transport protocol).
It could also make it simpler to extend, if we want to go in a dynamic or context-aware direction. And more searchable in the codebase to (e.g. if we need to `@Deprecate`) it.
What do you think?

thecoop commented:
I was going to use `NodeFeature` to check - have a method on DiscoveryNode or FeatureService that logically does `boolean nodeHasFeature(DiscoveryNode, NodeFeature)`, then the actual implementation is hidden from the caller.

thecoop committed:
Add some javadoc

thecoop committed:
Simplify the feature store

thecoop committed:
Add a couple of feature uses

thecoop commented:
I've added a couple of uses. I'm not happy with a few of the things here:

- I've changed NodeFeature to work in a similar way to settings - you just declare a static final variable somewhere, and just reference it. This means it's not enforced that you go through FeatureService to get a NodeFeature instance, which means you might skip registration. It also means that plugins have to declare the feature era, which external authors might not know what that is. Not great.
- Handling backwards compatibility is really tricky. I don't like the information leak from xpack plugin stuff directly into the historical versions list in FeatureService that 'these are the features across the whole of ES that are present'. Ideally we would have some kind of way for plugins to declare historical versions, conditional on whether they're present or not, but that's getting very complicated.

thecoop commented on server/src/main/java/org/elasticsearch/features/FeatureService.java:
```diff
@@ -0,0 +1,97 @@
+     */
+    private static final Map<Version, Set<String>> HISTORICAL_FEATURES = Map.ofEntries(
+        entry(Version.V_8_10_0, Set.of("elastic-connectors.templates")),
+        entry(Version.V_8_11_0, Set.of("esql.stats_node"))
```
This whole thing is a massive information leak from xpack. Also these plugins might not be installed!

ldematte replied:
Yes, that's not ideal. Maybe we can replicate the same schema in each of the plugins, building historical feature maps in each of them and registering them as regular features via PluginService - there should be no difference. It's more effort, but it will be more precise (if a plugin is not installed the feature is not there) and contain information better.

williamrandolph replied:
We could have a `FeatureProvidingPlugin` interface that could return a map of legacy Version-linked features and a set of new features. But once we're defining features in plugins, I think we need to add some kind of name spacing so that the string identifiers can't clash.

Node startup could refuse to proceed if two plugins try to provide the same feature.

ldematte replied:
Name spacing was also on my mind - I had it in my own take(s) at Features previously. One nice side effect is to be able to quickly identify a feature (set) owner. I don't know if we need it now, but it might be one of those things it would be difficult to introduce later.

thecoop replied:
That's a general thing though - node startup fails if anything provides two features with the same name.

I like the idea of namespacing plugin features, but is there any use case for moving features between plugins? That would cause lots of problems, with no way to affect the namespacing.

thecoop replied:
I've added historical feature registration. Because of its use in `DiscoveryNode`, it needs to be static - which is not great, but I don't think there's another way to do it

ldematte replied:
I understand your weariness: you have plugin instance register -> static registration method -> static historic feature set <- static read method <- discoverynode instance.

Timing works out correct but you dependent on it, without a good way of enforcing it.

I would like to revisit some sort of "static initialization is registration".
Would you think something like 
`static final NodeFeature ESQL_STATS_FEATURE = FeatureService.registerHistoricalFeature("esql.stats_node",  V_8_10_0);` could work?

It would create the `NodeFeature` with the correct era and call the current `FeatureService.registerHistoricalFeature` implementation (or just replace it).

For consistency, we could have 
`static final NodeFeature ESQL_STATS_FEATURE = FeatureService.registerFeature("esql.some_other_feature",  8);` too; that also creates the NodeFeature, and then proceed with the current `registerFeature` logic.

This way you give up the check on the pluginEra / declared era. But let me ask: is it important? 
My understanding is that in the current status we can just use the node version era, because plugins will have the same version; maybe in the future we will have relax it, but do we expect a cluster to load plugins from a different major? 
And even if we end up registering a feature for a past era, would it be bad?

I am pretty sure I could be missing something important thought :)

thecoop replied:
I don't think the whole `FeatureService` class can be static, as we need to be able to handle multiple `Node` instances per JVM in tests. The historical versions can be static, as that doesn't change, but the featureservice itself needs to be instanced to the node to allow different nodes in the same JVM to have different features.

thecoop replied:
I suppose there's no particular reason why we should limit the version for plugin historical features, if the plugin is updated and the feature is not...

ldematte replied:
> as we need to be able to handle multiple Node instances per JVM in tests 

That's a very good reason

thecoop commented:
There's an additional issue - a lot of tests create `DiscoveryNode` instances off their own bat. By default, they will have no features (apart from historical features). This means that when new features are added that you want to test, it will be easy to get tests that don't actually hit the case you want to test, as the feature to do so isn't there.

ldematte commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlStatsAction.java:
```diff
@@ -62,7 +64,7 @@ protected void resolveRequest(EsqlStatsRequest request, ClusterState clusterStat
         DiscoveryNode[] concreteNodes = Arrays.stream(nodesIds)
             .map(clusterState.nodes()::get)
-            .filter(n -> n.getVersion().onOrAfter(Version.V_8_11_0))
+            .filter(n -> n.hasFeature(ESQL_STATS_FEATURE))
```
This is neat :) I like it!

ldematte commented:
> This means it's not enforced that you go through FeatureService to get a NodeFeature instance, which means you might skip registration. It also means that plugins have to declare the feature era, which external authors might not know what that is. Not great.

This might be overcomplicated/crazy/not work, but maybe we can make a NodeFeature auto-registering, by reflection or if we want to avoid it, by adding a registration method:
```
static final NodeFeature ESQL_STATS_FEATURE = PluginService.registerNodeFeature("esql.stats_node", 8);
```
We can make NodeFeature an interface/internal so it can be created only by us.
`PluginService.registerNodeFeature` is static and will put NodeFeature in a (static) list (of features themselves or as registration functions/lambdas), and `Plugin.registerFeatures` will go through them, registering each of them. The (static) list can then be emptied/locked at the end. 

Personally, I prefer the style you have now with the `registerFeature.accept`, but it's true that having declaration and registration separated can bring problems. And having the static final FEATURE is so convenient at check time.

thecoop committed:
Add historical feature registration

thecoop committed:
Make historical feature reg idempotent

thecoop commented on server/src/main/java/org/elasticsearch/features/FeatureService.java:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        // make this method idempotent (so it can be called multiple times in the same JVM for tests)
+        if (HISTORICAL_FEATURES.getOrDefault(version, Set.of()).contains(feature.id())) {
```
This won't work in the case a `Node` is created & started, then a plugin is added and another `Node` is created - as the feature set will have been locked by then. You would have to register all the plugins before the first `Node` is created

ldematte replied:
This is a testing scenario or something that could really happen? It means that 2 `Node` instances must be created inside the same JVM, right?

thecoop replied:
Yup, it happens with some IT tests - but all plugins would be the same in that case

thecoop commented on x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorTemplateRegistryTests.java:
```diff
@@ -89,7 +89,7 @@ public void tearDown() throws Exception {
 
     public void testThatNonExistingComposableTemplatesAreAddedImmediately() throws Exception {
-        DiscoveryNode node = DiscoveryNodeUtils.create("node");
+        DiscoveryNode node = DiscoveryNodeUtils.builder("node").features(CONNECTOR_TEMPLATES_FEATURE).build();
```
So, this is an issue - the feature registration is done by the plugin, which is not loaded here. This means that no historical features are specified at all, so the feature needs to be explicitly added in tests. Not great.

Having a static registration method would fix historical features, but not new feature added after FeatureService is merged - those would always need to be explicitly specified whenever `DiscoveryNode` is used directly.

thecoop replied:
This would mean that every plugin that wants to auto-register a feature would have to go through the faff of using the java SPI stuff.

I'm not sure how much of an issue this is, given these uses are largely internal...

ldematte replied:
> would have to go through the faff of using the java SPI stuff.

Yeah, my thoughts exactly. I'm not sure is much of an issue either, you can consider it part of mocking -- setup tests with a partial/non complete infrastructure.

thecoop replied:
I've changed to use an SPI interface. I'm much happier with this, but we will still need a different mechanism for external plugins later

ldematte replied:
I might be biased as I've been struggling with SPIs all week :) but I liked the non-SPI approach more. 
I feel the problem of having to declare a feature and then register it separately (potentially forgetting to register it) now is a bit worse: you have to remember to add them to a different class (and to create it in the first place).

thecoop replied:
The issue is that the features need to be declared implicity, so that you can just use it in `DiscoveryNode` without changing anything. We also need to be able to control when features get registered, and potentially re-register them for tests etc. Hence the use of an interface implementation.

I think one class per plugin that is responsible for all feature registrations isn't too bad - it only needs to be set up once.

ldematte replied:
> I think one class per plugin that is responsible for all feature registrations isn't too bad - it only needs to be set up once.

That's true, we might be worrying for no reason

thecoop committed:
Change to an SPI interface to load versions

thecoop committed:
Allow historical versions to be re-calculated for tests

thecoop committed:
Tighten up on how and where historical features are specified

thecoop committed:
Fix npe

thecoop committed:
Merge branch 'main' into feature-registration

williamrandolph commented:
I like the way this is looking! The SPI approach seems very clean, and I think having the named features will eventually increase readability in the codebase.

Are we going to want to keep adding "historical features" as we release on-prem versions? There may be some cases in logging or error messages where we'll want to be able to tell self-managing users which artifact they need to download and install for up- or down-grading.

thecoop commented:
Historical features are only for features introduced prior to 8.11 (this will be enforced). All new features should just be added as a new feature without a version number

thecoop commented:
The real version is #100330
